File: USACO1/Chapter 3/3.2/Spinning Wheels/main.cpp
/*
ID: andreww7
PROG: spin
LANG: C++11
*/

#include <algorithm>
#include <array>
#include <bitset>
#include <cstdio>
#include <cstdarg>
#include <deque>
#include <fstream>
#include <iostream>
#include <iomanip>
#include <map> 
#include <math.h>
#include <queue>
#include <stdlib.h>
#include <string>
#include <sstream>
#include <vector>

using namespace std;

FILE *fin = fopen("spin.in", "r");
FILE *fout = fopen("spin.out", "w");

int gcd(int a, int b) {
	if (!b) return a;
	return gcd(b, a%b);
}

// The only use of lcm is in computing the upper bound for
// the number of seconds we have to simulate
int lcm1(int a, int b) {
	return a * b / gcd(a, b);
}

int lcm(int a, int b, int c, int d, int e) {
	return lcm1(a, lcm1(b, lcm1(c, lcm1(d, e))));
}

int main()
{
	int speed[5];
	pair<int, int> wedge[5][5];
	for (int i = 0; i < 5; ++i) {
		int w;
		fscanf(fin, "%d %d ", &speed[i], &w);
		for (int j = 0; j < w; ++j) {
			int a, b;
			fscanf(fin, "%d %d ", &a, &b);
			wedge[i][j] = make_pair(a, b);
		}
	}

    return 0;
}

File: USACO1/Chapter 3/3.2/Spinning Wheels/main.py
'''
ID: andreww7
LANG: PYTHON2
PROB: spin
'''
from fractions import gcd

def lcm(l):
	if len(l) == 2:
		return l[0]*l[1]/gcd(l[0], l[1])
	return lcm(l[0:1] + [lcm(l[1:])])

speed = []; wedge = [[] for i in range(5)]

def aligned():
	w = []
	for pair in wedge[0]: 
		w.append(pair)
	

def main():
	with open("spin.in", "r") as f:
		for i in range(5):
			l = list(map(int, next(f).split()))
			speed.append(l[0])
			for j in range(l[1]):
				wedge[i].append([l[2+2*j], l[2+2*j] + l[3+2*j]])
	f.close()
	M = lcm(speed)
	print M
	i = 0
	while (i < M):
		print i
		if aligned():
			break
		i += 1
		for j in range(5):
			for k in range(len(wedge[j])):
				wedge[j][k][0] = (wedge[j][k][0] + speed[j]) % 360
				wedge[j][k][1] = (wedge[j][k][1] + speed[j]) % 360

	with open("spin.out", "w") as output: 
		if i == M: 
			output.write('none\n')
			output.close()
			return
		output.write(str(i) + '\n')
		output.close()
				
main()